### PR TITLE
Tokenizer listeners optimizer

### PR DESCRIPTION
**Pull request by butschster**
| Q             | A
| ------------- | ---
| Bugfix?       | ❌
| Breaks BC?    | ❌
| New feature?  | ✔️

Tokenizer listener now can be cached. This is a very important feature, because it allows to speed up the application boot.
To use this feature, you should add `#[\Spiral\Tokenizer\Attribute\TargetClass::class]` or `#[\Spiral\Tokenizer\Attribute\TargetAttribute::class]` attribute to your listener and specify a target class.

For example

```php
#[TargetClass(class: Command::class)]
#[TargetAttribute(class: Route::class)]
final class RouteLocatorListener implements TokenizationListenerInterface
{
    ...
}
```

After adding the attribute, Spiral will automatically cache all found classes for the specified target class.
The cache will be stored in the file cache and will be used during the next application boot.

> **Note**:
> Use `TOKENIZER_CACHE_TARGETS` env variable to cache listener targets. Bu default, it's `false`

Here are some measures:

**Before caching:**
1.5448 sec.
**After:**
0.0032 sec.

### CODE REVIEW

**Review by msmakouz — APPROVED** [review]

File: src/Tokenizer/src/Attribute/TargetClass.php [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+#[\Attribute(\Attribute::TARGET_CLASS | \Attribute::IS_REPEATABLE), NamedArgumentConstructor]
+final class TargetClass
+{
+    /**
+     * @param class-string $class
+     * @param non-empty-string|null $scope
+     */
+    public function __construct(
+        public readonly string $class,
+        public readonly ?string $scope = null
```
Comment (roxblnfk) [inline src/Tokenizer/src/Attribute/TargetClass.php:22]:
```suggestion
        public readonly ?string $scope = null,
```

File: src/Tokenizer/src/Listener/AttributesParser.php [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+use Spiral\Tokenizer\Attribute\TargetClass;
+use Spiral\Tokenizer\TokenizationListenerInterface;
+
+/**
+ * @internal
+ */
+final class AttributesParser
+{
+    public function __construct(
+        private readonly ReaderInterface $reader
```
Comment (roxblnfk) [inline src/Tokenizer/src/Listener/AttributesParser.php:18]:
```suggestion
        private readonly ReaderInterface $reader,
```

File: src/Tokenizer/src/Listener/AttributesParser.php [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+{
+    public function __construct(
+        private readonly ReaderInterface $reader,
+    ) {
+    }
+
+    /**
+     * @return \Generator<ListenerDefinitionInterface>
+     */
+    public function parse(TokenizationListenerInterface $listener): \Generator
```
Comment (roxblnfk) [inline src/Tokenizer/src/Listener/AttributesParser.php:24]: Move to ClassesLocator

File: src/AnnotatedRoutes/src/RouteLocatorListener.php [hunk 4]
```diff
@@ -9,16 +9,18 @@
...
 use Spiral\Router\Annotation\Route;
 use Spiral\Router\Target\Action;
+use Spiral\Tokenizer\Attribute\TargetAttribute;
 use Spiral\Tokenizer\TokenizationListenerInterface;
 
+#[TargetAttribute(class: Route::class, useAnnotations: true)]
 final class RouteLocatorListener implements TokenizationListenerInterface
 {
-    /** @var array<array-key, array{0: ReflectionMethod, 1: Route}>  */
+    /** @var array<array-key, array{0: ReflectionMethod, 1: Route}> */
```
Comment (roxblnfk) [inline src/AnnotatedRoutes/src/RouteLocatorListener.php:18]:
```suggestion
    /** @var array<array-key, array{ReflectionMethod, Route}> */
```

File: src/Tokenizer/src/Attribute/AbstractTarget.php [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+namespace Spiral\Tokenizer\Attribute;
+
+abstract class AbstractTarget implements \Stringable
+{
+    /**
+     * @param class-string $class
+     * @param non-empty-string|null $scope
+     */
+    public function __construct(
+        public readonly string $class,
```
Comment (roxblnfk) [inline src/Tokenizer/src/Attribute/AbstractTarget.php]: Move to children

File: src/Prototype/src/Traits/PrototypeTrait.php [hunk 6]
```diff
@@ -11,6 +11,41 @@
...
+ * @property \Spiral\Storage\BucketInterface $storage
+ * @property \Spiral\Serializer\SerializerManager $serializer
+ * @property \Spiral\Validation\ValidationInterface $validator
+ * @property \Spiral\Views\ViewsInterface $views
+ * @property \Spiral\Auth\AuthScope $auth
+ * @property \Spiral\Auth\TokenStorageInterface $authTokens
+ * @property \Psr\SimpleCache\CacheInterface $cache
+ * @property \Spiral\Cache\CacheStorageProviderInterface $cacheManager
+ * @property \Spiral\Exceptions\ExceptionHandlerInterface $exceptionHandler
+ * @property \Spiral\App\SomeService\Client $service.client
```
Comment (roxblnfk) [inline src/Prototype/src/Traits/PrototypeTrait.php]:
```suggestion
```

**Review by roxblnfk — APPROVED** [review]

**Pull request merged**